aya-bpf: introduce argument coercion

willfindlay opened the pull request:
This PR introduces argument coercion for pt_regs and BTF program types and implements it for kprobes, uprobes, LSM programs, and BTF tracepoints.

You can see a [more concrete example of using this API here][1].

Features:
  - Uses a FromBtfArgs trait with blanket implementations to enable coercion to both primitive types _and_ raw pointers
  - No need for the user to perform any kind of manual casting on the BPF side
  - Similar logic for pt_regs programs, abstracting over the PT_REGS access

TODO:
  - Need to support architectures other than x86 for k/uprobe arguments (bindings need to be fixed)
  
While I was at it, I modified the Cargo.toml file of aya-bpf-macros to fix broken doctests. Perhaps we should add doctests to the CI workflow?

Closes #21

[1]: https://github.com/willfindlay/aya-lsm-args-example

alessandrod commented on bpf/aya-bpf/src/btf_args.rs:
```diff
@@ -0,0 +1,46 @@
+
+unsafe impl<T> FromBtfArgument for *const T {
+    unsafe fn from_argument(ctx: *const c_void, position: usize) -> *const T {
+        *(ctx as *const *const c_void).add(position * 8) as _
```
I'm not sure I understand, shouldn't it be `ctx.add((1 + position) * mem::size_of::<*const c_void>()) as _`?

willfindlay replied:
Good catch, I'll fix it

willfindlay replied:
@alessandrod I've updated this to be more semantically meaningful. While I was at it I just switched to use ptr.offset instead of ptr.add. Much simpler and more correct.

willfindlay commented:
@alessandrod Any thoughts on how this looks from an API perspective? Once I have the green light I will add supports for `pt_regs`-style contexts. Tracepoints are a bit trickier, since their contexts are actually just interpreted as a pointer to a tracepoint-specific C struct, so the API might diverge a bit there.

alessandrod commented:
It generally looks great! I think you should go ahead and try it with `pt_regs` as that will inform the design more. For example I think you'll need something like `FromBtfArgument` there too, so that might have to be renamed. 

Then I'd probably call it `ctx.arg(n)` instead of argument, for consistency with `std::env::args()`, `.arg()` from clap etc.

willfindlay commented:
@alessandrod I've added support for pt_regs probes. I have some example code over at https://github.com/willfindlay/aya-kprobe-arg-example demonstrating the API on a try_to_wake_up kprobe.

I'll definitely squash these commits before we land this, but I'm leaving the history mostly as-is for now to make things easier.

alessandrod commented on bpf/aya-bpf/src/args.rs:
```diff
@@ -0,0 +1,186 @@
+    /// memory. In particular, the value of `n` must not exceed the number of function
+    /// arguments. Moreover, `ctx` must be a valid pointer to a BTF context, and `T` must
+    /// be the right type for the given argument.
+    unsafe fn from_argument(ctx: *const c_void, n: isize) -> Self;
```
Why is `n: isize` and not `usize`? (here and in other places)

willfindlay replied:
isize is needed for `ptr::offset`

willfindlay replied:
other places is just for api consistency

willfindlay replied:
I suppose we could also accept a usize and just cast to an isize here though, up to you

alessandrod replied:
That's an implementation detail leaking out tho, we should take `usize` since you can't get negative arguments, and cast as `isize` in the impl. We don't allow large usize values anyway.

alessandrod commented on bpf/aya-bpf/src/args.rs:
```diff
@@ -0,0 +1,186 @@
+#[cfg(bpf_target_arch = "aarch64")]
+impl<T> FromPtRegs for *const T {
+    fn from_argument(ctx: &pt_regs, n: isize) -> Option<Self> {
+        if n <= 5 {
```
Shouldn't this be `if n <= 6`? (r0-r7)

willfindlay replied:
libbpf only defines PT_REGSPARM0 up to PT_REGS_PARM5, so I did the same here. See https://github.com/libbpf/libbpf/blob/17d7f04e7cb8cca664dfccd27c3848d105c56b5e/src/bpf_tracing.h#L200-L222. Are these extra registers documented somewhere?

alessandrod replied:
if we go the libbpf route, then it should be `n <= 4`. See https://en.wikipedia.org/wiki/Calling_convention

alessandrod commented on bpf/aya-bpf/src/programs/lsm.rs:
```diff
@@ -10,6 +10,50 @@ impl LsmContext {
         LsmContext { ctx }
     }
+
+    /// Coerces a `T` from the `n`th argument to the LSM hook where `n` starts at 0 and
```
I think this is a bit too technical, I'd rephrase it as:

```
/// Returns the `n`th argument passed to the LSM hook, starting from 0.
```

willfindlay replied:
Sure

alessandrod commented on bpf/aya-bpf/src/programs/probe.rs:
```diff
@@ -12,6 +18,52 @@ impl ProbeContext {
         }
     }
+
+    /// Coerces a `T` from the `n`th argument to the probed function where `n` starts at
```
And this as:

```
Returns the `n`th argument passed to the probed function, starting from 0.
```

willfindlay replied:
Sure

alessandrod commented on bpf/aya-bpf/src/programs/tp_btf.rs:
```diff
@@ -10,6 +10,43 @@ impl BtfTracePointContext {
         BtfTracePointContext { ctx }
     }
+
+    /// Coerces a `T` from the `n`th argument to the LSM hook where `n` starts at 0 and
```
```
Returns the `n`th argument passed to the tracepoint, starting from 0.
```

willfindlay commented:
@alessandrod I've applied your documentation nits, changed all signatures to accept a `usize` rather than an `isize`, and also changed the maximum allowed `n` for aarch64 and armv7. I mostly followed https://man7.org/linux/man-pages/man2/syscall.2.html (see the section labelled Architecture calling convention).  Hopefully, I haven't gotten anything wrong, since we're kind of going off the reservation w.r.t. libbpf.

P.S. I'm noticing a pattern where PRs I work on after 2am all have off-by-one errors, so maybe it's time to rethink my sleeping habits :)

willfindlay commented:
Ah clippy doesn't like casting for offsets, I'll refactor.

alessandrod commented on bpf/aya-bpf/src/args.rs:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    fn from_retval(ctx: &pt_regs) -> Option<Self> {
+        ctx.uregs().map(|regs| regs[14] as _)
```
I think this should be `regs[0]`?

willfindlay replied:
Ah whoops good catch

alessandrod commented on bpf/aya-bpf/src/args.rs:
```diff
@@ -0,0 +1,196 @@
+#[cfg(bpf_target_arch = "armv7")]
+impl<T> FromPtRegs for *const T {
+    fn from_argument(ctx: &pt_regs, n: usize) -> Option<Self> {
+        // See https://man7.org/linux/man-pages/man2/syscall.2.html under
```
I don't think this link is too relevant - kprobes can execute in non-syscall contexts, and uprobes use `ProbeContext` too

willfindlay replied:
Right but this at least appears to be the calling convention they are referencing in libbpf so I figured this would be the best place to consult for pt_regs-related accesses beyond the scope of what libbpf does (just the first five arguments). Alternatively, we could just truncate everything to 5 arguments like libbpf does and call it a day.

alessandrod commented on bpf/aya-bpf/src/args.rs:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    fn from_retval(ctx: &pt_regs) -> Option<Self> {
+        ctx.regs().map(|regs| regs[30] as _)
```
I think this should be `regs[0]` too

alessandrod commented on bpf/aya-bpf/src/programs/probe.rs:
```diff
@@ -12,6 +18,50 @@ impl ProbeContext {
+        T::from_argument(unsafe { &*self.regs }, n)
+    }
+
+    /// Coerces a `T` from the return value of the probed function.
```
I'd phrase this as:

```
Returns the return value of the probed function.
```

alessandrod reviewed (changes requested):
I'm starting to think that we should probably not try to implement `ProbeContext::ret()` at all. The registers used for return values depend on the return type (eg int vs float), and can actually span multiple registers.

alessandrod commented on bpf/aya-bpf/src/args.rs:
```diff
@@ -0,0 +1,196 @@
+
+            fn from_retval(ctx: &pt_regs) -> Option<Self> {
+                ctx.regs().map(|regs| regs[30] as _)
+            }
```
this should be `regs[0]`

alessandrod commented on bpf/aya-bpf/src/args.rs:
```diff
@@ -0,0 +1,196 @@
+            }
+
+            fn from_retval(ctx: &pt_regs) -> Option<Self> {
+                ctx.uregs().map(|regs| regs[14] as _)
```
This should be `regs[0]`

alessandrod reviewed (changes requested):
Oops we missed a couple. This should be the last one 🤞

willfindlay commented:
> Oops we missed a couple. This should be the last one 🤞

Ah whoops. I'll have some time to fix this later tonight!

willfindlay commented:
@alessandrod should be fixed now

alessandrod commented:
Great work! 🙏
